nucliadb/nucliadb/writer/back_pressure.py

```diff
@@ -0,0 +1,436 @@
+        with self._lock:
+            self._cache[key] = data
+
+
+_cache = BackPressureCache()
+
+
+@contextlib.contextmanager
+def cached_back_pressure(*args):
+    """
+    Context manager that handles the caching of the try again in time so that
+    we don't recompute try again times if we have already applied back pressure.
+    """
+
+    cache_key = "-".join(args)
+
+    data: Optional[BackPressureData] = _cache.get(cache_key)
+    if data is not None:
+        raise HTTPException(
+            status_code=429,
+            detail={
+                "message": f"Too many messages pending to ingest. Retry after {data.try_after}",
+                "try_after": data.try_after.timestamp(),
+                "back_pressure_type": data.type,
+            },
+        )
+    try:
+        yield
+    except BackPressureException as exc:
+        _cache.set(cache_key, exc.data)
+        raise HTTPException(
+            status_code=429,
+            detail={
+                "message": f"Too many messages pending to ingest. Retry after {exc.data.try_after}",
+                "try_after": exc.data.try_after.timestamp(),
+                "back_pressure_type": exc.data.type,
+            },
+        )
+
+
+class Materializer:
+    """
+    Singleton class that will run in the background gatheringthe different stats to
+    apply back pressure and materializing it in memory. This allows us to do
+    stale-reads when checking if back pressure is needed for a particular
+    request - thus not slowing it down.
+    """
+
+    def __init__(
+        self,
+        nats_manager: NatsConnectionManager,
+        indexing_check_interval: int = 30,
+        ingest_check_interval: int = 30,
+    ):
+        self.nats_manager = nats_manager
+        self.processing_http_client = ProcessingHTTPClient()
+
+        self.indexing_check_interval = indexing_check_interval
+        self.ingest_check_interval = ingest_check_interval
+
+        self.ingest_pending: int = 0
+        self.indexing_pending: dict[str, int] = {}
+
+        self._tasks: list[asyncio.Task] = []
+        self._running = False
+
+    async def start(self):
+        self._tasks.append(asyncio.create_task(self._get_indexing_pending_task()))
+        self._tasks.append(asyncio.create_task(self._get_ingest_pending_task()))
+        self._running = True
+
+    async def stop(self):
+        for task in self._tasks:
+            task.cancel()
+        self._tasks.clear()
+        await self.processing_http_client.close()
+        self._running = False
+
+    @property
+    def running(self) -> bool:
+        return self._running
+
+    @alru_cache(maxsize=1024, ttl=60)
+    async def get_processing_pending(self, kbid: str) -> int:
+        # TODO: turn it into a settigns-configurabe ttl cache !!!!
+
+        """
+        We won't materialize the pending messages for every KB, but we will cache the result some time.
+        """
+        response = await self.processing_http_client.stats(kbid=kbid)
```

Comment:
Would 0.5 as a timeout be low enough?

Right now if there's an error, I just return 0 so that the request doesn't fail. I make sure not to cache the value in that case.

Should I send a setry event in that case so we get alerted?

Reply:
1/2 second is probably a fine timeout.